File: apps/api/app/schemas/ai_content.py
"""Pydantic schemas for AI-generated content payloads."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class ManifestData(BaseModel):
    """Manifest metadata for a single AI content generation."""

    activity_type: str = Field(..., max_length=128)
    title: str = Field(..., max_length=512)
    item_count: int = Field(..., ge=0)
    has_audio: bool = Field(default=False)
    has_passage: bool = Field(default=False)
    difficulty: str | None = Field(default=None, max_length=32)
    language: str = Field(default="en", max_length=64)
    created_by: str | None = Field(default=None, max_length=255)
    created_at: datetime | None = Field(default=None)


class AIContentCreate(BaseModel):
    """Payload for creating a new AI content entry (manifest + content)."""

    manifest: ManifestData
    content: dict


class AIContentCreateResponse(BaseModel):
    """Response after creating AI content."""

    content_id: str
    storage_path: str


class ManifestRead(ManifestData):
    """Manifest returned in list responses, includes content_id."""

    content_id: str


class AIContentRead(BaseModel):
    """Full AI content response (manifest + content)."""

    content_id: str
    manifest: ManifestData
    content: dict


class AudioUploadResponse(BaseModel):
    """Response after uploading a single audio file."""

    filename: str
    storage_path: str
    size: int


class BatchAudioResponse(BaseModel):
    """Response after batch audio upload."""

    uploaded: list[AudioUploadResponse]
    failed: list[str] = Field(default_factory=list)

File: apps/admin-panel/src/lib/aiContent.ts
import { ApiClient, apiClient } from './api';
import { buildAuthHeaders } from './http';
import { buildApiUrl } from '../config/environment';

// ---------------------------------------------------------------------------
// Types (matching backend schemas in ai_content.py)
// ---------------------------------------------------------------------------

export interface ManifestRead {
  content_id: string;
  activity_type: string;
  title: string;
  item_count: number;
  has_audio: boolean;
  has_passage: boolean;
  difficulty: string | null;
  language: string;
  created_by: string | null;
  created_at: string | null;
}

export interface AIContentRead {
  content_id: string;
  manifest: Omit<ManifestRead, 'content_id'>;
  content: Record<string, unknown>;
}

export interface AudioUploadResponse {
  filename: string;
  storage_path: string;
  size: number;
}

export interface BatchAudioResponse {
  uploaded: AudioUploadResponse[];
  failed: string[];
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

/**
 * List all AI content manifests for a book.
 */
export const listAIContent = (
  bookId: number,
  token: string,
  tokenType: string = 'Bearer',
  client: ApiClient = apiClient
): Promise<ManifestRead[]> =>
  client.get<ManifestRead[]>(`/books/${bookId}/ai-content/`, {
    headers: buildAuthHeaders(token, tokenType),
  });

/**
 * Get full AI content (manifest + content.json) for a specific content_id.
 */
export const getAIContent = (
  bookId: number,
  contentId: string,
  token: string,
  tokenType: string = 'Bearer',
  client: ApiClient = apiClient
): Promise<AIContentRead> =>
  client.get<AIContentRead>(`/books/${bookId}/ai-content/${contentId}`, {
    headers: buildAuthHeaders(token, tokenType),
  });

/**
 * Hard-delete an AI content generation (manifest, content, audio).
 */
export const deleteAIContent = (
  bookId: number,
  contentId: string,
  token: string,
  tokenType: string = 'Bearer',
  client: ApiClient = apiClient
): Promise<{ content_id: string; objects_removed: number }> =>
  client.delete<{ content_id: string; objects_removed: number }>(
    `/books/${bookId}/ai-content/${contentId}`,
    undefined,
    { headers: buildAuthHeaders(token, tokenType) }
  );

/**
 * Build a full URL for streaming an audio file (uses auth header at fetch time).
 */
export const getAIContentAudioUrl = (
  bookId: number,
  contentId: string,
  filename: string
): string =>
  buildApiUrl(`/books/${bookId}/ai-content/${contentId}/audio/${encodeURIComponent(filename)}`);
